Added section on testing python packages.

ChrisBarker-NOAA opened the pull request:
As per #345, I've  added stuff about testing python packages.

I put it in the testing page, and added a link to it from the step by step instructions.

This could use some more proofreading and definitely more content, but it's a start.

jakirkham commented:
Could you please take a look, @artPlusPlus?

jakirkham commented on src/testing.rst:
```diff
@@ -29,8 +34,125 @@ On Windows, use the ``exist`` command. See below for an example.
+.. code-block:: yaml
+
+    test:
+      imports
```
Needs a `:` after.

artPlusPlus commented on src/recipe.rst:
```diff
@@ -27,7 +27,7 @@ Step-by-step instructions
 2. Within your forked copy, generate a new folder in the recipes subdirectory and copy `meta.yml <https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml>`_ from the example directory. Please leave the example directory unchanged!
 3. Edit the copied recipe (meta.yml) as you need it
 4. Generate the ssh 256 key like described in the example recipe using the `openssl` tool. As an alternative you can also go to the package description on `PyPi <https://pypi.org>`_ where you can directly copy the ssh 256 key from.
```
I think `ssh 256` should be `sha256`.

artPlusPlus commented on src/recipe.rst:
```diff
@@ -26,7 +26,7 @@ Step-by-step instructions
 2. Within your forked copy, generate a new folder in the recipes subdirectory and copy `meta.yml <https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml>`_ from the example directory. Please leave the example directory unchanged!
 3. Edit the copied recipe (meta.yml) as you need it
-4. Generate the ssh 256 key like described in the example recipe using the `openssl` tool. As an alternative you can also go to the package description on `PyPi <https://pypi.org>`_ where you can directly copy the ssh 256 key from.
+4. Generate the sha256 key like described in the example recipe using the `openssl` tool. As an alternative you can also go to the package description on `PyPi <https://pypi.org>`_ where you can directly copy the ssh 256 key from.
```
I promise I'm not trying to be a jerk, but there's one more towards the end of the line: ssh 256 -> sha256

ChrisBarker-NOAA commented:
On Wed, Mar 15, 2017 at 2:47 AM, Matt Robinson <notifications@github.com>
wrote:

> I promise I'm not trying to be a jerk, but there's one more towards the
> end of the line: ssh 256 -> sha256
>
> aarrgg!

I think I got them all now....

-CHB


-- 

Christopher Barker, Ph.D.
Oceanographer

Emergency Response Division
NOAA/NOS/OR&R            (206) 526-6959   voice
7600 Sand Point Way NE   (206) 526-6329   fax
Seattle, WA  98115       (206) 526-6317   main reception

Chris.Barker@noaa.gov

jakirkham commented on src/testing.rst:
```diff
@@ -29,8 +34,125 @@ On Windows, use the ``exist`` command. See below for an example.
+
+This makes it very hard to run tests that are not installed with the package.
+
+**NOTE** if anyone has good ideas as to how to do that, please put it here!
```
Recently had to do this with a bunch of packages actually. 😄

Here's an [example]( https://github.com/conda-forge/mplview-feedstock/blob/85bf7f2f17b6da46551bf4f76f2868d16d4d692d/recipe/meta.yaml#L28-L36 ) of one. Can provide more examples if you would like. Key thing is if this is a package on PyPI, one will need to ensure the tests are included in their package. This can be done by adding them to one's [MANIFEST.in]( https://github.com/conda-forge/mplview-feedstock/blob/85bf7f2f17b6da46551bf4f76f2868d16d4d692d/recipe/meta.yaml#L28-L36 ). Feel free to pull from this info liberally. No need for attribution.

ChrisBarker-NOAA replied:
I see this:

```
test:
  source_files:
    - tests
```

so where does conda build look for "tests"???

As I understand it, the source tarball is NOT unpacked in the test environment. The case I'm trying to catch here is when there are tests in the source tarball, but not in the installed package.

-CHB

jakirkham replied:
Right, that is this case. The `source_files` section says grab this from the unpacked tarball and copy them (recursively) into the test environment. This way we can use these selected files and directories in the test environment; even though, they were never installed into the prefix. Here's a link to the [docs]( https://conda.io/docs/building/meta-yaml.html#source-files ) though it is a bit terse. The conda-build 2 release [blog entry]( https://www.continuum.io/blog/developer-blog/whats-old-and-new-conda-build ) is a bit more detailed. Does that help clarify it?

jakirkham replied:
> so where does conda build look for "tests"???

To be clear, from the source directory (`SRC_DIR`).

jakirkham commented on src/testing.rst:
```diff
@@ -51,9 +174,10 @@ the root repository directory and run, the ``scripts/run_docker_build.sh`` scrip
 No. A recipe does not have to run all of a package's unit or integration tests.
 Sometimes, this might even be impossible due to timeouts on CI services.
 Just test sufficiently to know that the package is in good working order.
+
```
nit: Maybe this is just due to Windows formatting, in which case don't worry about it. Do we need this extra newline?

jakirkham commented on src/testing.rst:
```diff
@@ -29,8 +34,125 @@ On Windows, use the ``exist`` command. See below for an example.
+nose Tests
+...........
+
+Someone please add a description and example for running nose on an installed package.
```
Basically the same store, but add `nose` to `requires` and change the command to `nosetests package_name`.

jakirkham commented:
Thanks for doing this, @ChrisBarker-NOAA! This is really great.

Added some info for the yet to be filled out sections. Hope that helps.

I'm wondering if we should actually move the Docker test stuff to a different section. This is a healthy section on Python testing, but the two aren't really related. That said, don't feel obliged to handle this. I or someone else can do some cleanup if you don't have time. Though adding an issue to track this wouldn't hurt (assuming you agree with my conclusions).

ChrisBarker-NOAA commented:
Ping! any reason not to merge this?

BTW -- I updated the icons to higher res, and added an SVG version (Inkscape) I shoul dhave done that in a different branch, but I totally forgot I had this PR in waiting...

jakirkham commented:
So I had a couple of comments above, is there a plan to address them? 😕

If there are testing examples that you would like to see, but are still unsure how to demonstrate them after our discussion above, would recommend cutting those sections from the docs and simply raising issues on this repo to fill them out. However this is merely my preference.

pelson commented on src/testing.rst:
```diff
@@ -1,10 +1,15 @@
 All recipes need tests. Here are some tips, tricks, and justifications.
+How you shold test depends on the type of package (python, c-lib,
+command-line tool, ... ), and what tests are available for that package.
+But every conda pacakge should have at least *some* tests.
```
should -> must

pelson commented on src/testing.rst:
```diff
@@ -29,8 +34,125 @@ On Windows, use the ``exist`` command. See below for an example.
+including nose, pytest, etc.
+
+Also, some packages install the tests with the package, and thus they can be
+run in place. while others keep the tests with the source code, and thus can
```
period -> comma

pelson commented on src/testing.rst:
```diff
@@ -29,8 +34,125 @@ On Windows, use the ``exist`` command. See below for an example.
+      commands:
+        - py.test --pyargs package_name
+
+nose Tests
```
Let's simply remove it. We can't document *everything* 😉

pelson commented on src/testing.rst:
```diff
@@ -29,8 +34,125 @@ On Windows, use the ``exist`` command. See below for an example.
 
+Testing Python Packages
+-----------------------
+
```
The first thing we should do is to reference the conda-build docs on this https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html#test-section.

I'm keen to document what we need to for conda-forge, but for everything else try to push upstream to conda-build.

Heading that I think we should have:

 * Testing imports
 * Running unit tests (pytest)
 * Testing CLIs

ChrisBarker-NOAA replied:
added link to conda docs.

pelson commented:
@ChrisBarker-NOAA - looks like this has stalled a little.
Since it has merge conflicts, I'll close the PR for now. Please feel free to re-open when it is ready to go.

Thanks!

jakirkham commented on src/testing.rst:
```diff
@@ -28,8 +33,125 @@ On Windows, use the ``exist`` command. See below for an example.
+
 Run Docker Tests Locally for Feedstock
---------------------------------------
+.......................................
```
Not sure if this was intentional, but don't think Sphinx will like this.

ChrisBarker-NOAA replied:
this was intentional, and it renders fine with sphinx -- makes a subsubsection...

jakirkham replied:
Interesting, was not aware of that. Thanks for teaching me something new. :)

jakirkham replied:
Should we do the same for `staged-recipes` below?

ChrisBarker-NOAA replied:
looking at it rendered, the headings are a bit wonky.

I'll clean that up.

jakirkham commented:
@pelson, anything else needed here?

ChrisBarker-NOAA commented on src/recipe.rst:
```diff
@@ -26,28 +29,28 @@ Your final recipe should have no comments (unless they're actually relevant to t
-1. Fork the `example recipes <https://github.com/conda-forge/staged-recipes/tree/master/recipes>`_
-2. Within your forked copy, generate a new folder in the recipes subdirectory and copy `meta.yml <https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml>`_ from the example directory. Please leave the example directory unchanged!
+1. Fork `example recipes <https://github.com/conda-forge/staged-recipes/tree/master/recipes>`_
+2. Within your forked copy, generate a new folder in the recipes subdirectory and copy `meta.yml <https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml>`_ file. from the example directory. Please leave the example directory unchanged!
```
fixed.

ChrisBarker-NOAA commented:
OK -- let's put this one to bed!

jakirkham commented:
Thanks for sticking with this, @ChrisBarker-NOAA. A welcome improvement indeed. :)
